src/engineer/engineer.py

```diff
@@ -0,0 +1,229 @@
+        years.sort()
+
+        output_test_arrays: DDict[int, DDict[int, Dict[str, xr.Dataset]]] = \
+            defaultdict(lambda: defaultdict(dict))
+
+        xy_test, min_test_date = self.stratify_xy(ds, years[0], target_variable, 1, pred_months,
+                                                  expected_length, True)
+
+        train_dates = ds.time.values <= np.datetime64(str(min_test_date))
+        train_ds = ds.isel(time=train_dates)
+
+        if xy_test is not None:
+            output_test_arrays[years[0]][1] = xy_test
+
+        for year in years:
+            for month in range(1, 13):
+                if year > years[0] or month > 1:
+                    # prevents the initial test set from being recalculated
+                    xy_test, _ = self.stratify_xy(ds, year, target_variable, month, pred_months,
+                                                  expected_length)
+                    if xy_test is not None:
+                        output_test_arrays[year][month] = xy_test
+
+        return train_ds, output_test_arrays
+
+    @staticmethod
+    def stratify_xy(ds: xr.Dataset,
+                    year: int,
+                    target_variable: str,
+                    target_month: int,
+                    pred_months: int,
+                    expected_length: Optional[int],
+                    return_min_date: bool = False,
+                    ) -> Tuple[Optional[Dict[str, xr.Dataset]], Optional[date]]:
+
+        print(f'Generating data for year: {year}, target month: {target_month}')
+
+        max_date = date(year, target_month, calendar.monthrange(year, target_month)[-1])
+        mx_year, mx_month, max_train_date = minus_months(year, target_month, diff_months=1)
+        _, _, min_date = minus_months(mx_year, mx_month, diff_months=pred_months)
+
+        min_date_np = np.datetime64(str(min_date))
+        max_date_np = np.datetime64(str(max_date))
+        max_train_date_np = np.datetime64(str(max_train_date))
+
+        print(f'Max date: {str(max_date)}, max train date: {str(max_train_date)}, '
+              f'min train date: {str(min_date)}')
+
+        if return_min_date:
+            output_min_date: Optional[date] = min_date
+        else:
+            output_min_date = None
+
+        x = ((ds.time.values > min_date_np) & (ds.time.values <= max_train_date_np))
+        y = ((ds.time.values > max_train_date_np) & (ds.time.values <= max_date_np))
+
+        if sum(y) != 1:
+            print(f'Wrong number of test y values! Expected 1, got {sum(y)}; returning None')
+            return None, output_min_date
+
+        if expected_length is not None:
+            if sum(x) != expected_length:
+                print(f'Wrong number of test x values! Got {sum(x)} Returning None')
+
+                return None, output_min_date
+
+        x_dataset = ds.isel(time=x)
+        y_dataset = ds.isel(time=y)[target_variable].to_dataset()
```

Comment:
We should give the to_dataset() a name in case the `DataArray` is unnamed ( it will throw an error if the `DataArray` is unnamed).

http://xarray.pydata.org/en/stable/generated/xarray.DataArray.to_dataset.html

```
ds.isel(time=y)[target_variable].to_dataset(name=target_variable)
```

Although it might not be necessary because we've already selected that `target_variable`.

IT's just an error i have come across before

Reply:
Updated!